lib/phoenix_live_view/tag_engine.ex

```diff
@@ -767,15 +803,85 @@ defmodule Phoenix.LiveView.TagEngine do
         |> set_root_on_not_tag()
         |> push_tag({:tag, name, new_attrs, new_meta})
         |> handle_tag_and_attrs(name, new_attrs, ">", to_location(new_meta))
+        |> continue(tokens)
     end
   end
 
-  defp handle_token({:close, :tag, name, tag_meta} = token, state) do
+  defp handle_token([{:close, :tag, name, tag_meta} = token | tokens], state) do
     {{:tag, _name, _attrs, tag_open_meta}, state} = pop_tag!(state, token)
 
     state
     |> update_subengine(:handle_text, [to_location(tag_meta), "</#{name}>"])
     |> handle_special_expr(tag_open_meta)
+    |> continue(tokens)
+  end
+
+  defp handle_token([], state), do: state
+
+  defp handle_macro_component(
+         [{:tag, _name, _attrs, %{macro_component: module_string} = tag_meta} | _] = tokens,
+         state
+       ) do
+    Macro.Env.required?(state.caller, Phoenix.Component) ||
+      raise ArgumentError,
+            "macro components are only supported in modules that `use Phoenix.Component`"
+
+    module = validate_module!(module_string, tag_meta, state)
+
+    {ast, rest} =
+      case Phoenix.Component.MacroComponent.build_ast(tokens) do
+        {:ok, ast, rest} -> {ast, rest}
+        {:error, message, meta} -> raise_syntax_error!(message, meta, state)
+      end
+
+    try do
+      module.transform(ast, %{file: state.file, line: tag_meta.line, env: state.caller})
+    rescue
+      e in ArgumentError ->
+        raise_syntax_error!(
+          Exception.message(e),
+          tag_meta,
+          state
+        )
+    else
+      {:ok, new_ast} ->
+        new_tokens = Phoenix.Component.MacroComponent.ast_to_tokens(new_ast)
+        continue(state, new_tokens ++ rest)
+
+      {:ok, new_ast, data} ->
+        Module.put_attribute(state.caller.module, :__macro_components__, {module, data})
+        new_tokens = Phoenix.Component.MacroComponent.ast_to_tokens(new_ast)
+        continue(state, new_tokens ++ rest)
+
+      other ->
+        raise ArgumentError,
+              "a macro component must return {:ok, ast} or {:ok, ast, data}, got: #{inspect(other)}"
+    end
+  end
+
+  defp validate_module!(module_string, tag_meta, state) do
+    module =
+      case Code.eval_string(module_string, [], state.caller) do
+        {mod, _} when is_atom(mod) ->
+          mod
+
+        _ ->
+          raise_syntax_error!(
+            "the given macro component #{inspect(module_string)} is not a valid module",
+            tag_meta,
+            state
+          )
+      end
+
+    if not (Code.ensure_loaded?(module) and function_exported?(module, :transform, 2)) do
```

Comment:
```suggestion
    if not (Code.ensure_compiled?(module) and function_exported?(module, :transform, 2)) do
```

With the new lazy loading, I need ensure_compiled here. Do I need to adjust the other calls as well, or is it ensure_loaded guaranteed to work inside a mix compiler (as everything is definitely compiled then)?

Reply:
Yes, it is needed at compile-time. The benefit of v1.19 is exactly that this behaviour is now consistent. Before it could work on certain cases by accident.


```suggestion
    # We need the module, let the compiler raise if not available
    Code.ensure_compiled!(module)

    if not function_exported?(module, :transform, 2) do
```